activerecord/lib/active_record/autosave_association.rb

```diff
@@ -456,9 +456,17 @@ def save_has_one_association(reflection)
 
       # If the record is new or it has changed, returns true.
       def record_changed?(reflection, record, key)
-        record.new_record? ||
-          (record.has_attribute?(reflection.foreign_key) && record[reflection.foreign_key] != key) ||
-          record.will_save_change_to_attribute?(reflection.foreign_key)
+        return true if record.new_record?
+        return true if _association_foreign_key_changed?(reflection, record, key)
+        return true if record.will_save_change_to_attribute?(reflection.foreign_key)
+
+        false
+      end
+
+      def _association_foreign_key_changed?(reflection, record, key)
```

Comment:
The leading `_` is less worth, since it is private (people can't call it directly), and the four words method name is less likely to conflict with user-defined method names, either is fine through.

Reply:
Agree, removed the leading underscore